src/dir_contents.ml

```diff
@@ -374,246 +193,461 @@ let build_coq_modules_map (d : _ Dir_with_dune.t) ~dir ~modules =
       Lib_name.Map.add map (Dune_file.Coq.best_name coq) modules
     | _ -> map)
 
-type result0_here = {
-  t : t;
-  (* [rules] includes rules for subdirectories too *)
-  rules : Rules.t option;
-  subdirs : t Path.Build.Map.t;
-}
+module rec Load : sig
+  val get : Super_context.t -> dir:Path.Build.t -> t
+  val gen_rules : Super_context.t -> dir:Path.Build.t -> gen_rules_result
+end = struct
+  let virtual_modules sctx vlib =
+    let info = Lib.info vlib in
+    let lib_modules =
+      match Option.value_exn (Lib_info.virtual_ info) with
+      | External lib_modules -> lib_modules
+      | Local ->
+        let src_dir =
+          Lib_info.src_dir info
+          |> Path.as_in_build_dir_exn
+        in
+        let t = Load.get sctx ~dir:src_dir in
+        modules_of_library t ~name:(Lib.name vlib)
+    in
+    let existing_virtual_modules =
+      Lib_modules.virtual_modules lib_modules
+      |> Module.Name.Map.keys
+      |> Module.Name.Set.of_list
+    in
+    let allow_new_public_modules =
+      Lib_modules.wrapped lib_modules
+      |> Wrapped.to_bool
+      |> not
+    in
+    { Modules_field_evaluator.Implementation.
+      existing_virtual_modules
+    ; allow_new_public_modules
+    }
 
-type result0 =
-  | See_above of Path.Build.t
-  | Here of result0_here
+  let make_modules sctx (d : _ Dir_with_dune.t) ~modules =
+    let scope = d.scope in
+    let libs, exes =
+      List.filter_partition_map d.data ~f:(fun stanza ->
+        match (stanza : Stanza.t) with
+        | Library lib ->
+          let src_dir = d.ctx_dir in
+          let resolved = lazy (
```

Comment:
This uses a builtin lazy and as we've previously seen, they are not playing nice with memoization in general.

I wonder if this use is all right though. It seems all right because the lazy doesn't escape the stack frame it's created in (as far as I can tell).

Maybe there should be a new type of lazy that enforces this restriction: only allow the lazy to be forced from the same memo stack frame where it's created. (or we could use Memo.Lazy, but that has higher overhead)

Reply:
I reworked the code to avoid the lazy. As a bonus, it makes the code a tad simpler.